datalad/cmd.py

```diff
@@ -98,7 +98,208 @@ def _cleanup_output(stream, std):
+        try:
+            lgr.log(8, "Start running %r", cmd)
+            process = subprocess.Popen(
+                cmd,
+                # from PY37 onwards
+                #capture_output=proc_output,
+                stdout=subprocess.PIPE if proc_out else None,
+                stderr=subprocess.PIPE if proc_out else None,
+                shell=False,
+                cwd=self.cwd,
+                env=self.env,
+                stdin=stdin,
+                # intermediate reports are never decoded anyways
+                # from PY37 onwards
+                #text=False,
+                universal_newlines=False,
+            )
+        except Exception as e:
+            lgr.log(11, "Failed to start %r%r: %s" %
+                    (cmd,
+                     (" under %r" % self.cwd) if self.cwd else '',
+                     exc_str(e)))
+            raise
+
+        try:
+            out = [b'', b'']
+            last_idx = [0, 0]
+            pout = None
+            # make sure to run this loop at least once, even if the
+            # process is already dead
+            keep_going = True
+            while process.poll() is None or keep_going:
+                try:
+                    # get a chunk of output for the specific period
+                    # of time
+                    pout = process.communicate(
+                        timeout=self._poll_period,
+                    )
+                    # we know the process ended at this point
+                    keep_going = False
+                except subprocess.TimeoutExpired as poll:
+                    # this will always be the full report so far,
+                    # not just an output increment
+                    pout = (poll.stdout, poll.stderr)
+
+                if proc_out is not None:
+                    for i, (o, proc) in enumerate(zip(pout, proc_out)):
+                        if proc is None:
+                            if o is not None:
+                                out[i] += o
+                            # no index update needed, can never change
+                            continue
+                        # current length of output
+                        len_o = len(o) if o else 0
+                        # anything new in the output?
+                        if len_o - last_idx[i] > 0:
+                            # engage output processed
+                            processed, unprocessed_len = proc(o[last_idx[i]:])
+                            # only keep the processed output
+                            out[i] += processed
+                            # record point to which we processed the output
+                            # subtract number of bytes reported as unprocessed
+                            last_idx[i] = len_o - unprocessed_len
+
+            # obtain exit code
+            status = process.poll()
+
+            # decode bytes to string
+            out = tuple(o.decode('utf-8') if o else '' for o in out)
```

Comment:
I suppose 'utf-8' is consistent with what we do in `ensure_unicode` (before we fall back to all the guesswork), though I think it'd be better to use `locale.getpreferredencoding(False)`.

Reply:
This comment still applies to spots in latest revision (5a758c303) that use `.decode('utf-8')`.

Reply:
Thanks for the reminder, done now.